File: ios/Classes/document/BookmarkManagerImpl.swift
//
//  Copyright © 2024-2026 PSPDFKit GmbH. All rights reserved.
//
//  THIS SOURCE CODE AND ANY ACCOMPANYING DOCUMENTATION ARE PROTECTED BY INTERNATIONAL COPYRIGHT LAW
//  AND MAY NOT BE RESOLD OR REDISTRIBUTED. USAGE IS BOUND TO THE PSPDFKIT LICENSE AGREEMENT.
//  UNAUTHORIZED REPRODUCTION OR DISTRIBUTION IS SUBJECT TO CIVIL AND CRIMINAL PENALTIES.
//  This notice may not be removed from this file.
//

import Foundation
import PSPDFKit
import Flutter

/// Implementation of the BookmarkManagerApi for managing PDF bookmarks on iOS.
///
/// This class bridges the Flutter bookmark API to the PSPDFKit iOS SDK's
/// bookmark functionality. It supports adding, removing, updating, and querying bookmarks.
@objc(BookmarkManagerImpl)
public class BookmarkManagerImpl: NSObject, BookmarkManagerApi {

    // MARK: - Properties
    private var document: Document?
    private let documentId: String
    private let binaryMessenger: FlutterBinaryMessenger
    private let queue = DispatchQueue(label: "com.pspdfkit.bookmark-manager", qos: .userInitiated)

    // MARK: - Initialization
    init(documentId: String, binaryMessenger: FlutterBinaryMessenger) {
        self.documentId = documentId
        self.binaryMessenger = binaryMessenger
        super.init()
    }

    // MARK: - BookmarkManagerApi Implementation

    func initialize(documentId: String) throws {
        guard let doc = AnnotationManagerImpl.getDocument(withId: documentId) else {
            throw BookmarkManagerError.documentNotFound(documentId)
        }
        self.document = doc
    }

    func getBookmarks(completion: @escaping (Result<[Bookmark], Error>) -> Void) {
        queue.async {
            guard let document = self.document else {
                DispatchQueue.main.async {
                    completion(.failure(BookmarkManagerError.documentNotInitialized))
                }
                return
            }

            let nativeBookmarks = document.bookmarks
            let bookmarks = nativeBookmarks.map { self.convertToFlutterBookmark($0) }

            DispatchQueue.main.async {
                completion(.success(bookmarks))
            }
        }
    }

    func addBookmark(bookmark: Bookmark, completion: @escaping (Result<Bookmark, Error>) -> Void) {
        queue.async {
            guard let document = self.document else {
                DispatchQueue.main.async {
                    completion(.failure(BookmarkManagerError.documentNotInitialized))
                }
                return
            }

            do {
                let nativeBookmark = try self.convertToNativeBookmark(bookmark, document: document)
                document.bookmarkManager?.addBookmark(nativeBookmark)

                let resultBookmark = self.convertToFlutterBookmark(nativeBookmark)

                DispatchQueue.main.async {
                    completion(.success(resultBookmark))
                }
            } catch {
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }

    func removeBookmark(bookmark: Bookmark, completion: @escaping (Result<Bool, Error>) -> Void) {
        queue.async {
            guard let document = self.document else {
                DispatchQueue.main.async {
                    completion(.failure(BookmarkManagerError.documentNotInitialized))
                }
                return
            }

            // Find the bookmark to remove
            let pageIndex = self.parsePageIndexFromAction(bookmark.actionJson)
            let bookmarkToRemove = document.bookmarks.first { nativeBookmark in
                if let action = nativeBookmark.action as? GoToAction {
                    return Int(action.pageIndex) == pageIndex
                }
                // Fallback to name matching
                return bookmark.name != nil && nativeBookmark.name == bookmark.name
            }

            if let bookmarkToRemove = bookmarkToRemove {
                document.bookmarkManager?.removeBookmark(bookmarkToRemove)
                DispatchQueue.main.async {
                    completion(.success(true))
                }
            } else {
                DispatchQueue.main.async {
                    completion(.success(false))
                }
            }
        }
    }

    func updateBookmark(bookmark: Bookmark, completion: @escaping (Result<Bool, Error>) -> Void) {
        // PSPDFKit iOS doesn't have a direct update method, so we remove and re-add
        removeBookmark(bookmark: bookmark) { [weak self] removeResult in
            guard let self = self else {
                completion(.failure(BookmarkManagerError.documentNotInitialized))
                return
            }

            switch removeResult {
            case .success:
                self.addBookmark(bookmark: bookmark) { addResult in
                    switch addResult {
                    case .success:
                        completion(.success(true))
                    case .failure(let error):
                        completion(.failure(error))
                    }
                }
            case .failure:
                // If remove failed (bookmark didn't exist), just add it
                self.addBookmark(bookmark: bookmark) { addResult in
                    switch addResult {
                    case .success:
                        completion(.success(true))
                    case .failure(let error):
                        completion(.failure(error))
                    }
                }
            }
        }
    }

    func getBookmarksForPage(pageIndex: Int64, completion: @escaping (Result<[Bookmark], Error>) -> Void) {
        queue.async {
            guard let document = self.document else {
                DispatchQueue.main.async {
                    completion(.failure(BookmarkManagerError.documentNotInitialized))
                }
                return
            }

            let filteredBookmarks = document.bookmarks.filter { nativeBookmark in
                if let action = nativeBookmark.action as? GoToAction {
                    return Int64(action.pageIndex) == pageIndex
                }
                return false
            }.map { self.convertToFlutterBookmark($0) }

            DispatchQueue.main.async {
                completion(.success(filteredBookmarks))
            }
        }
    }

    func hasBookmarkForPage(pageIndex: Int64, completion: @escaping (Result<Bool, Error>) -> Void) {
        getBookmarksForPage(pageIndex: pageIndex) { result in
            switch result {
            case .success(let bookmarks):
                completion(.success(!bookmarks.isEmpty))
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }

    // MARK: - Helper Methods

    /// Convert a PSPDFKit Bookmark to Flutter Bookmark.
    private func convertToFlutterBookmark(_ nativeBookmark: PSPDFKit.Bookmark) -> Bookmark {
        var actionJson: String? = nil

        if let action = nativeBookmark.action as? GoToAction {
            let actionDict: [String: Any] = [
                "type": "goTo",
                "pageIndex": action.pageIndex,
                "destinationType": "fitPage"
            ]
            if let jsonData = try? JSONSerialization.data(withJSONObject: actionDict),
               let jsonString = String(data: jsonData, encoding: .utf8) {
                actionJson = jsonString
            }
        }

        return Bookmark(
            pdfBookmarkId: nativeBookmark.identifier,
            name: nativeBookmark.name,
            actionJson: actionJson
        )
    }

    /// Convert Flutter Bookmark to a PSPDFKit Bookmark.
    private func convertToNativeBookmark(_ bookmark: Bookmark, document: Document) throws -> PSPDFKit.Bookmark {
        let pageIndex = parsePageIndexFromAction(bookmark.actionJson) ?? 0
        let action = GoToAction(pageIndex: PageIndex(pageIndex))

        let name = bookmark.name ?? "Page \(pageIndex + 1)"

        // Use existing identifier if provided, otherwise generate a new one
        let identifier = bookmark.pdfBookmarkId ?? UUID().uuidString

        // Generate sortKey based on current bookmark count for ordering
        let sortKey = NSNumber(value: document.bookmarks.count)

        return PSPDFKit.Bookmark(identifier: identifier, action: action, name: name, sortKey: sortKey)
    }

    /// Parse the page index from the action JSON.
    private func parsePageIndexFromAction(_ actionJson: String?) -> Int? {
        guard let actionJson = actionJson,
              let jsonData = actionJson.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any],
              let type = json["type"] as? String,
              type == "goTo",
              let pageIndex = json["pageIndex"] as? Int else {
            return nil
        }
        return pageIndex
    }

    // MARK: - Plugin Registration

    @objc public func register(binaryMessenger: FlutterBinaryMessenger) {
        BookmarkManagerApiSetup.setUp(
            binaryMessenger: binaryMessenger,
            api: self,
            messageChannelSuffix: "\(documentId)_bookmark_manager"
        )
    }

    @objc public func unregister(binaryMessenger: FlutterBinaryMessenger) {
        BookmarkManagerApiSetup.setUp(
            binaryMessenger: binaryMessenger,
            api: nil,
            messageChannelSuffix: "\(documentId)_bookmark_manager"
        )
    }

    // MARK: - Factory Methods

    @objc public static func create(documentId: String, binaryMessenger: FlutterBinaryMessenger) -> BookmarkManagerImpl {
        let manager = BookmarkManagerImpl(documentId: documentId, binaryMessenger: binaryMessenger)
        manager.register(binaryMessenger: binaryMessenger)
        return manager
    }

    @objc public static func createAndInitialize(document: Document, binaryMessenger: FlutterBinaryMessenger) -> BookmarkManagerImpl {
        let documentId = document.uid ?? UUID().uuidString

        // Create and register bookmark manager
        let manager = BookmarkManagerImpl(documentId: documentId, binaryMessenger: binaryMessenger)
        manager.register(binaryMessenger: binaryMessenger)

        // Initialize with document reference
        do {
            try manager.initialize(documentId: documentId)
        } catch {
            print("BookmarkManager: Failed to initialize with document \(documentId): \(error)")
        }

        return manager
    }
}

// MARK: - Error Definitions
enum BookmarkManagerError: LocalizedError {
    case documentNotFound(String)
    case documentNotInitialized
    case invalidJSON(String)

    var errorDescription: String? {
        switch self {
        case .documentNotFound(let id):
            return "Document not found: \(id)"
        case .documentNotInitialized:
            return "Document not initialized"
        case .invalidJSON(let message):
            return "Invalid JSON: \(message)"
        }
    }
}

File: android/src/main/java/com/pspdfkit/flutter/pspdfkit/document/BookmarkManagerImpl.kt
/*
 * Copyright © 2024-2026 PSPDFKit GmbH. All rights reserved.
 * <p>
 * THIS SOURCE CODE AND ANY ACCOMPANYING DOCUMENTATION ARE PROTECTED BY INTERNATIONAL COPYRIGHT LAW
 * AND MAY NOT BE RESOLD OR REDISTRIBUTED. USAGE IS BOUND TO THE PSPDFKIT LICENSE AGREEMENT.
 * UNAUTHORIZED REPRODUCTION OR DISTRIBUTION IS SUBJECT TO CIVIL AND CRIMINAL PENALTIES.
 * This notice may not be removed from this file.
 */

package com.pspdfkit.flutter.pspdfkit.document

import android.util.Log
import com.pspdfkit.bookmarks.Bookmark as NativeBookmark
import com.pspdfkit.document.PdfDocument
import com.pspdfkit.flutter.pspdfkit.api.Bookmark
import com.pspdfkit.flutter.pspdfkit.api.BookmarkManagerApi
import com.pspdfkit.flutter.pspdfkit.api.NutrientApiError
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.schedulers.Schedulers
import org.json.JSONObject

/**
 * Implementation of the BookmarkManagerApi for managing PDF bookmarks on Android.
 *
 * This class bridges the Flutter bookmark API to the PSPDFKit Android SDK's
 * BookmarkProvider. It supports adding, removing, updating, and querying bookmarks.
 */
class BookmarkManagerImpl : BookmarkManagerApi {

    companion object {
        private const val TAG = "BookmarkManagerImpl"
    }

    private var documentId: String? = null
    private val disposables = CompositeDisposable()

    private val pdfDocument: PdfDocument?
        get() {
            val docId = documentId ?: return null
            return FlutterPdfDocument.documentInstances[docId]?.pdfDocument
        }

    override fun initialize(documentId: String) {
        this.documentId = documentId
        Log.d(TAG, "Initialized BookmarkManager for document: $documentId")
    }

    override fun getBookmarks(callback: (Result<List<Bookmark>>) -> Unit) {
        val document = pdfDocument
        if (document == null) {
            callback(Result.failure(NutrientApiError("Document not found", "Document with ID $documentId not found")))
            return
        }

        val disposable = document.bookmarkProvider.getBookmarksAsync()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { bookmarks ->
                    val bookmarkList = bookmarks.map { convertToBookmark(it) }
                    callback(Result.success(bookmarkList))
                },
                { error ->
                    Log.e(TAG, "Error getting bookmarks", error)
                    callback(Result.failure(NutrientApiError("Failed to get bookmarks", error.message ?: "")))
                }
            )
        disposables.add(disposable)
    }

    override fun addBookmark(bookmark: Bookmark, callback: (Result<Bookmark>) -> Unit) {
        val document = pdfDocument
        if (document == null) {
            callback(Result.failure(NutrientApiError("Document not found", "Document with ID $documentId not found")))
            return
        }

        try {
            val nativeBookmark = convertToNativeBookmark(bookmark)

            val disposable = document.bookmarkProvider.addBookmarkAsync(nativeBookmark)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    {
                        // Return the bookmark with the ID assigned
                        val resultBookmark = convertToBookmark(nativeBookmark)
                        callback(Result.success(resultBookmark))
                    },
                    { error ->
                        Log.e(TAG, "Error adding bookmark", error)
                        callback(Result.failure(NutrientApiError("Failed to add bookmark", error.message ?: "")))
                    }
                )
            disposables.add(disposable)
        } catch (e: Exception) {
            Log.e(TAG, "Error creating bookmark", e)
            callback(Result.failure(NutrientApiError("Failed to create bookmark", e.message ?: "")))
        }
    }

    override fun removeBookmark(bookmark: Bookmark, callback: (Result<Boolean>) -> Unit) {
        val document = pdfDocument
        if (document == null) {
            callback(Result.failure(NutrientApiError("Document not found", "Document with ID $documentId not found")))
            return
        }

        val disposable = document.bookmarkProvider.getBookmarksAsync()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { bookmarks ->
                    // Find the bookmark to remove by matching page index or name
                    val pageIndex = parsePageIndexFromAction(bookmark.actionJson)
                    val bookmarkToRemove = bookmarks.find { existingBookmark ->
                        // Match by page index
                        if (pageIndex != null && existingBookmark.pageIndex == pageIndex) {
                            true
                        } else {
                            // Fallback to name matching
                            bookmark.name != null && existingBookmark.name == bookmark.name
                        }
                    }

                    if (bookmarkToRemove != null) {
                        val removeDisposable = document.bookmarkProvider.removeBookmarkAsync(bookmarkToRemove)
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe(
                                {
                                    callback(Result.success(true))
                                },
                                { error ->
                                    Log.e(TAG, "Error removing bookmark", error)
                                    callback(Result.failure(NutrientApiError("Failed to remove bookmark", error.message ?: "")))
                                }
                            )
                        disposables.add(removeDisposable)
                    } else {
                        callback(Result.success(false))
                    }
                },
                { error ->
                    Log.e(TAG, "Error finding bookmark to remove", error)
                    callback(Result.failure(NutrientApiError("Failed to find bookmark", error.message ?: "")))
                }
            )
        disposables.add(disposable)
    }

    override fun updateBookmark(bookmark: Bookmark, callback: (Result<Boolean>) -> Unit) {
        val document = pdfDocument
        if (document == null) {
            callback(Result.failure(NutrientApiError("Document not found", "Document with ID $documentId not found")))
            return
        }

        // PSPDFKit Android doesn't have a direct update method, so we remove and re-add
        removeBookmark(bookmark) { removeResult ->
            if (removeResult.isSuccess) {
                addBookmark(bookmark) { addResult ->
                    if (addResult.isSuccess) {
                        callback(Result.success(true))
                    } else {
                        callback(Result.failure(addResult.exceptionOrNull() ?: Exception("Failed to add updated bookmark")))
                    }
                }
            } else {
                // If remove failed (bookmark didn't exist), just add it
                addBookmark(bookmark) { addResult ->
                    if (addResult.isSuccess) {
                        callback(Result.success(true))
                    } else {
                        callback(Result.failure(addResult.exceptionOrNull() ?: Exception("Failed to add bookmark")))
                    }
                }
            }
        }
    }

    override fun getBookmarksForPage(pageIndex: Long, callback: (Result<List<Bookmark>>) -> Unit) {
        val document = pdfDocument
        if (document == null) {
            callback(Result.failure(NutrientApiError("Document not found", "Document with ID $documentId not found")))
            return
        }

        val disposable = document.bookmarkProvider.getBookmarksAsync()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { bookmarks ->
                    val filteredBookmarks = bookmarks.filter { nativeBookmark ->
                        nativeBookmark.pageIndex == pageIndex.toInt()
                    }.map { convertToBookmark(it) }
                    callback(Result.success(filteredBookmarks))
                },
                { error ->
                    Log.e(TAG, "Error getting bookmarks for page", error)
                    callback(Result.failure(NutrientApiError("Failed to get bookmarks", error.message ?: "")))
                }
            )
        disposables.add(disposable)
    }

    override fun hasBookmarkForPage(pageIndex: Long, callback: (Result<Boolean>) -> Unit) {
        getBookmarksForPage(pageIndex) { result ->
            if (result.isSuccess) {
                val bookmarks = result.getOrNull() ?: emptyList()
                callback(Result.success(bookmarks.isNotEmpty()))
            } else {
                callback(Result.failure(result.exceptionOrNull() ?: Exception("Failed to check bookmark")))
            }
        }
    }

    /**
     * Convert a PSPDFKit NativeBookmark to Bookmark for Flutter.
     */
    private fun convertToBookmark(nativeBookmark: NativeBookmark): Bookmark {
        // Android SDK Bookmark uses pageIndex directly, not action
        val pageIndex = nativeBookmark.pageIndex
        val actionJson = if (pageIndex != null) {
            JSONObject().apply {
                put("type", "goTo")
                put("pageIndex", pageIndex)
                put("destinationType", "fitPage")
            }.toString()
        } else {
            null
        }

        return Bookmark(
            pdfBookmarkId = nativeBookmark.uuid, // Use UUID from Android SDK
            name = nativeBookmark.name,
            actionJson = actionJson
        )
    }

    /**
     * Convert Bookmark from Flutter to a PSPDFKit NativeBookmark.
     */
    private fun convertToNativeBookmark(bookmark: Bookmark): NativeBookmark {
        val pageIndex = parsePageIndexFromAction(bookmark.actionJson) ?: 0
        val name = bookmark.name ?: "Page ${pageIndex + 1}"

        // Android SDK Bookmark constructor takes name and pageIndex directly
        return NativeBookmark(name, pageIndex)
    }

    /**
     * Parse the page index from the action JSON.
     */
    private fun parsePageIndexFromAction(actionJson: String?): Int? {
        if (actionJson == null) return null

        return try {
            val json = JSONObject(actionJson)
            if (json.optString("type") == "goTo") {
                json.optInt("pageIndex", -1).takeIf { it >= 0 }
            } else {
                null
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error parsing action JSON", e)
            null
        }
    }

    /**
     * Clean up resources.
     */
    fun dispose() {
        disposables.clear()
        documentId = null
    }
}
